### ISSUE TITLE
mdd data access?

### ISSUE THREAD

**Issue by KonstantinDjairo** [dictlab/mdict-cpp#13]
if the output of a lookup made in a mdx file is something like that (a link) :
``` shell
@@@LINK=みち【道・路・途】
```

how should we lookup the actual content in the mdd file?
is there any function to do so?

**Comment by terasum** [comment]
As far as I know, you should find the word at mdx file, as your example, you may handle the link word 'みち【道・路・途】' as a key word to find in mdx file.

| |
ChenQuan
|
|
***@***.***
|
---- Replied Message ----
| From | ***@***.***> |
| Date | 12/27/2024 12:01 |
| To | dictlab/mdict-cpp ***@***.***> |
| Cc | Subscribed ***@***.***> |
| Subject | [dictlab/mdict-cpp] mdd data access? (Issue #13) |

if the output of a lookup made in a mdx file is something like that (a link) :

@@@LINK=みち【道・路・途】

how should we lookup the actual content in the mdd file?
is there any function to do so?

**Comment by KonstantinDjairo** [comment]
you're right, i should pass that same string to the lookup function again, but without the "@@@LINK="

thanks!

**Comment by KonstantinDjairo** [comment]
@terasum how do we access data from the mdd , like images and things like that? the output of the mdx is a html with a link to the data contained within the mdd, but i cant think of a way of actually getting the real data, like displaying an image for example

**Comment by terasum** [comment]
@KonstantinDjairo just like mdx, when you get the data use `lookup` or other methods, you will get the binary data, just save as  a image file (.jpg, .png, etc.) then you can use the image. I usually encode the binary data as base64 string, and use the data-url protocol to show the image.

Take a look at https://github.com/terasum/js-mdict, the example folder show you some code:
```
npx tsx ./example/oale8-mdd-example.ts

NOTE: the mdd's definition is base64 encoded bytes, 
if your target is css/js content, please decode base64 and get the original text
if your target is images, you can use dataurl to show the images

{
  keyText: '\\Logo.jpg',
  definition: '/9j/4AAQSkZJRgABAgAAAQABAAD//gAEKgD/4gIcSUNDX1BST0ZJTEUAAQEAAAIMbGNtcwIQ...'
 }
```

**Comment by KonstantinDjairo** [comment]
@terasum while i can lookup data in the mdx, i couldn't do the same for the mdd, it throws a segfault and also the output of the definition isnt base64:

```shell
hashirama@lisp-machine ~/dev/haskell/mdict-cpp/build/bin$ ./mydict  /mnt/Data/Japanese_Resources/languages-study\(japanesAndRussian\)/dicts/updated_ones/大辞泉/2024-09-11_21-19-49/大辞泉202304.mdd srej.css
arg count: 3
executable file name: ./mydict
/mnt/Data/Japanese_Resources/languages-study(japanesAndRussian)/dicts/updated_ones/大辞泉/2024-09-11_21-19-49/大辞泉202304.mdd
init cost time: 0ms

lookup cost time: 1 ms
key list length: 7
key list[0] word: 5c0041007500640069006f002e0070006e006700:0
key list[len-1] word: 5c007300720065006a002e006a007300:14044145
key list[len-1] def: 6c6574207372656a5f746869734247203d202223303037326336222c0d0a202020207372656a5f6f6e546162496e646578203d20302c0d0a202020207372656a5f636c69636b6564203d2066616c73652c0d0a202020207372656a5f6e6565644a756d70203d2066616c73653b0d0a0d0a2f2a6372656469743a206b6172782a2f0d0a66756e6374696f6e207372656a5f64287429207b0d0a202020207661722079203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a20202020666f72202865203d20303b2065203c20792e6c656e6774683b20652b2b29207b0d0a2020202020202020795b655d2e636c6173734c6973742e72656d6f7665282261637469766522293b0d0a202020207d0d0a20202020766172206c203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a6322293b0d0a0d0a20202020766172206e6577496e646578203d20742e69642e737562737472696e67283429202d20310d0a20202020696620287372656a5f636c69636b6564202626207372656a5f6f6e546162496e646578203d3d206e6577496e64657829207b0d0a2020202020202020666f72202865203d20303b2065203c206c2e6c656e6774683b20652b2b29207b0d0a2020202020202020202020206c5b655d2e7365744174747269627574652822636c617373222c20227372656a7922290d0a20202020202020207d0d0a20202020202020207372656a5f636c69636b6564203d2066616c73650d0a202020207d20656c7365207b0d0a20202020202020207372656a5f6f6e546162496e646578203d206e6577496e6465780d0a2020202020202020666f72202865203d20303b2065203c206c2e6c656e6774683b20652b2b29207b0d0a202020202020202020202020746162203d206c5b655d3b0d0a202020202020202020202020696620287372656a5f6f6e546162496e646578203d3d3d206529207b0d0a202020202020202020202020202020207461622e636c6173734c6973742e616464282261637469766522293b0d0a202020202020202020202020202020207461622e636c6173734c6973742e72656d6f766528226c6174656e7422293b0d0a2020202020202020202020207d20656c7365207b0d0a202020202020202020202020202020207461622e636c6173734c6973742e61646428226c6174656e7422293b0d0a202020202020202020202020202020207461622e636c6173734c6973742e72656d6f7665282261637469766522293b0d0a2020202020202020202020207d0d0a20202020202020207d0d0a2020202020202020742e636c6173734c6973742e616464282761637469766527293b0d0a20202020202020207372656a5f636c69636b6564203d20747275653b0d0a2020202020202020696620287372656a5f6e6565644a756d70292020747279207b2077696e646f772e6c6f636174696f6e2e68617368203d20742e6765744174747269627574652822696422293b207d206361746368286572726f7229207b7d0d0a202020207d0d0a7d0d0a0d0a66756e6374696f6e207372656a5f70287429207b0d0a20202020766172207469746c6573203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a2020202078203d20287372656a5f6f6e546162496e6465782d3129202520287469746c65732e6c656e677468293b0d0a2020202069662028783c30290d0a202020202020202078203d20287469746c65732e6c656e6774682b78293b0d0a202020207372656a5f64287469746c65735b785d290d0a7d0d0a0d0a66756e6374696f6e207372656a5f6e287429207b0d0a20202020766172207469746c6573203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a2020202078203d20287372656a5f6f6e546162496e6465782b3129202520287469746c65732e6c656e677468293b0d0a202020207372656a5f64287469746c65735b785d290d0a7d3b0d0a0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a205354415254204d4458205350454349464943204a415641534352495054202a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a0d0a
Segmentation fault
```

i'm trying to lookup the css

**Comment by terasum** [comment]
This issue usually occurs at some CJK dictionary, can you send me the original mdd file, so that I can test this? thank you!




---- Replied Message ----
| From | ***@***.***> |
| Date | 03/13/2025 11:33 |
| To | ***@***.***> |
| Cc | Quan ***@***.***> ,
***@***.***> |
| Subject | Re: [dictlab/mdict-cpp] mdd data access? (Issue #13) |

@terasum while i can lookup data in the mdx, i couldn't do the same for the mdd, it throws a segfault and also the output of the definition isnt base64:

***@***.*** ~/dev/haskell/mdict-cpp/build/bin$ ./mydict  /mnt/Data/Japanese_Resources/languages-study\(japanesAndRussian\)/dicts/updated_ones/大辞泉/2024-09-11_21-19-49/大辞泉202304.mdd srej.css
arg count: 3
executable file name: ./mydict
/mnt/Data/Japanese_Resources/languages-study(japanesAndRussian)/dicts/updated_ones/大辞泉/2024-09-11_21-19-49/大辞泉202304.mdd
init cost time: 0ms

lookup cost time: 1 ms
key list length: 7
key list[0] word: 5c0041007500640069006f002e0070006e006700:0
key list[len-1] word: 5c007300720065006a002e006a007300:14044145
key list[len-1] def: 6c6574207372656a5f746869734247203d202223303037326336222c0d0a202020207372656a5f6f6e546162496e646578203d20302c0d0a202020207372656a5f636c69636b6564203d2066616c73652c0d0a202020207372656a5f6e6565644a756d70203d2066616c73653b0d0a0d0a2f2a6372656469743a206b6172782a2f0d0a66756e6374696f6e207372656a5f64287429207b0d0a202020207661722079203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a20202020666f72202865203d20303b2065203c20792e6c656e6774683b20652b2b29207b0d0a2020202020202020795b655d2e636c6173734c6973742e72656d6f7665282261637469766522293b0d0a202020207d0d0a20202020766172206c203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a6322293b0d0a0d0a20202020766172206e6577496e646578203d20742e69642e737562737472696e67283429202d20310d0a20202020696620287372656a5f636c69636b6564202626207372656a5f6f6e546162496e646578203d3d206e6577496e64657829207b0d0a2020202020202020666f72202865203d20303b2065203c206c2e6c656e6774683b20652b2b29207b0d0a2020202020202020202020206c5b655d2e7365744174747269627574652822636c617373222c20227372656a7922290d0a20202020202020207d0d0a20202020202020207372656a5f636c69636b6564203d2066616c73650d0a202020207d20656c7365207b0d0a20202020202020207372656a5f6f6e546162496e646578203d206e6577496e6465780d0a2020202020202020666f72202865203d20303b2065203c206c2e6c656e6774683b20652b2b29207b0d0a202020202020202020202020746162203d206c5b655d3b0d0a202020202020202020202020696620287372656a5f6f6e546162496e646578203d3d3d206529207b0d0a202020202020202020202020202020207461622e636c6173734c6973742e616464282261637469766522293b0d0a202020202020202020202020202020207461622e636c6173734c6973742e72656d6f766528226c6174656e7422293b0d0a2020202020202020202020207d20656c7365207b0d0a202020202020202020202020202020207461622e636c6173734c6973742e61646428226c6174656e7422293b0d0a202020202020202020202020202020207461622e636c6173734c6973742e72656d6f7665282261637469766522293b0d0a2020202020202020202020207d0d0a20202020202020207d0d0a2020202020202020742e636c6173734c6973742e616464282761637469766527293b0d0a20202020202020207372656a5f636c69636b6564203d20747275653b0d0a2020202020202020696620287372656a5f6e6565644a756d70292020747279207b2077696e646f772e6c6f636174696f6e2e68617368203d20742e6765744174747269627574652822696422293b207d206361746368286572726f7229207b7d0d0a202020207d0d0a7d0d0a0d0a66756e6374696f6e207372656a5f70287429207b0d0a20202020766172207469746c6573203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a2020202078203d20287372656a5f6f6e546162496e6465782d3129202520287469746c65732e6c656e677468293b0d0a2020202069662028783c30290d0a202020202020202078203d20287469746c65732e6c656e6774682b78293b0d0a202020207372656a5f64287469746c65735b785d290d0a7d0d0a0d0a66756e6374696f6e207372656a5f6e287429207b0d0a20202020766172207469746c6573203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a2020202078203d20287372656a5f6f6e546162496e6465782b3129202520287469746c65732e6c656e677468293b0d0a202020207372656a5f64287469746c65735b785d290d0a7d3b0d0a0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a205354415254204d4458205350454349464943204a415641534352495054202a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a0d0a
Segmentation fault

i'm trying to lookup the css

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

KonstantinDjairo left a comment (dictlab/mdict-cpp#13)

@terasum while i can lookup data in the mdx, i couldn't do the same for the mdd, it throws a segfault and also the output of the definition isnt base64:

***@***.*** ~/dev/haskell/mdict-cpp/build/bin$ ./mydict  /mnt/Data/Japanese_Resources/languages-study\(japanesAndRussian\)/dicts/updated_ones/大辞泉/2024-09-11_21-19-49/大辞泉202304.mdd srej.css
arg count: 3
executable file name: ./mydict
/mnt/Data/Japanese_Resources/languages-study(japanesAndRussian)/dicts/updated_ones/大辞泉/2024-09-11_21-19-49/大辞泉202304.mdd
init cost time: 0ms

lookup cost time: 1 ms
key list length: 7
key list[0] word: 5c0041007500640069006f002e0070006e006700:0
key list[len-1] word: 5c007300720065006a002e006a007300:14044145
key list[len-1] def: 6c6574207372656a5f746869734247203d202223303037326336222c0d0a202020207372656a5f6f6e546162496e646578203d20302c0d0a202020207372656a5f636c69636b6564203d2066616c73652c0d0a202020207372656a5f6e6565644a756d70203d2066616c73653b0d0a0d0a2f2a6372656469743a206b6172782a2f0d0a66756e6374696f6e207372656a5f64287429207b0d0a202020207661722079203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a20202020666f72202865203d20303b2065203c20792e6c656e6774683b20652b2b29207b0d0a2020202020202020795b655d2e636c6173734c6973742e72656d6f7665282261637469766522293b0d0a202020207d0d0a20202020766172206c203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a6322293b0d0a0d0a20202020766172206e6577496e646578203d20742e69642e737562737472696e67283429202d20310d0a20202020696620287372656a5f636c69636b6564202626207372656a5f6f6e546162496e646578203d3d206e6577496e64657829207b0d0a2020202020202020666f72202865203d20303b2065203c206c2e6c656e6774683b20652b2b29207b0d0a2020202020202020202020206c5b655d2e7365744174747269627574652822636c617373222c20227372656a7922290d0a20202020202020207d0d0a20202020202020207372656a5f636c69636b6564203d2066616c73650d0a202020207d20656c7365207b0d0a20202020202020207372656a5f6f6e546162496e646578203d206e6577496e6465780d0a2020202020202020666f72202865203d20303b2065203c206c2e6c656e6774683b20652b2b29207b0d0a202020202020202020202020746162203d206c5b655d3b0d0a202020202020202020202020696620287372656a5f6f6e546162496e646578203d3d3d206529207b0d0a202020202020202020202020202020207461622e636c6173734c6973742e616464282261637469766522293b0d0a202020202020202020202020202020207461622e636c6173734c6973742e72656d6f766528226c6174656e7422293b0d0a2020202020202020202020207d20656c7365207b0d0a202020202020202020202020202020207461622e636c6173734c6973742e61646428226c6174656e7422293b0d0a202020202020202020202020202020207461622e636c6173734c6973742e72656d6f7665282261637469766522293b0d0a2020202020202020202020207d0d0a20202020202020207d0d0a2020202020202020742e636c6173734c6973742e616464282761637469766527293b0d0a20202020202020207372656a5f636c69636b6564203d20747275653b0d0a2020202020202020696620287372656a5f6e6565644a756d70292020747279207b2077696e646f772e6c6f636174696f6e2e68617368203d20742e6765744174747269627574652822696422293b207d206361746368286572726f7229207b7d0d0a202020207d0d0a7d0d0a0d0a66756e6374696f6e207372656a5f70287429207b0d0a20202020766172207469746c6573203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a2020202078203d20287372656a5f6f6e546162496e6465782d3129202520287469746c65732e6c656e677468293b0d0a2020202069662028783c30290d0a202020202020202078203d20287469746c65732e6c656e6774682b78293b0d0a202020207372656a5f64287469746c65735b785d290d0a7d0d0a0d0a66756e6374696f6e207372656a5f6e287429207b0d0a20202020766172207469746c6573203d20646f63756d656e742e676574456c656d656e747342795461674e616d6528227372656a7422293b0d0a2020202078203d20287372656a5f6f6e546162496e6465782b3129202520287469746c65732e6c656e677468293b0d0a202020207372656a5f64287469746c65735b785d290d0a7d3b0d0a0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a205354415254204d4458205350454349464943204a415641534352495054202a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a2f2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2a2f0d0a0d0a
Segmentation fault

i'm trying to lookup the css

**Comment by terasum** [comment]
@KonstantinDjairo And have you tried search the `/srej.css` as the search key? In some case, you need add `/` prefix to find the resource.

**Comment by KonstantinDjairo** [comment]
yes, i did tried using slashes and also backslashes, but only got segfaults...

**Comment by KonstantinDjairo** [comment]
> This issue usually occurs at some CJK dictionary, can you send me the original mdd file, so that I can test this? thank you!


here's the link to the files:
https://easyupload.io/m/uu89ug

**Comment by terasum** [comment]
@KonstantinDjairo hi, have you tried `\\srej.css`?

I used `js-mdict` library can get the resource:

```
$ npm install -g js-mdict@6.0.7
$ js-mdict ~/Downloads/uu89ug_folder/大辞泉202304.mdd \\srej.css 
dGFibGUuc3JlansgbWFyZ2luOjAgYXV0bztib3JkZXItY29sbGFwc2U6Y29sbGFwc2U7Ym9yZGVyLXN0eWxlOmhpZGRlbiB9DQp0...(total: 3976.97265625 KB)
```

**Comment by KonstantinDjairo** [comment]
> [@KonstantinDjairo](https://github.com/KonstantinDjairo) hi, have you tried `\\srej.css`?
> 
> I used `js-mdict` library can get the resource:
> 
> ```
> $ npm install -g js-mdict@6.0.7
> $ js-mdict ~/Downloads/uu89ug_folder/大辞泉202304.mdd \\srej.css 
> dGFibGUuc3JlansgbWFyZ2luOjAgYXV0bztib3JkZXItY29sbGFwc2U6Y29sbGFwc2U7Ym9yZGVyLXN0eWxlOmhpZGRlbiB9DQp0...(total: 3976.97265625 KB)
> ```

the mdict_test.cc in this repo gives a segfault instead of showing this result , but im glad it worked on the js-mdict version, maybe we can look at the js code and try to figure out what's wrong with the example shown in this repo...

**Comment by KonstantinDjairo** [comment]
updates:
it seems that mdict-cpp outputs the mdd data in hexadecimal format, but it can be converted to text and then we obtain a string encoded in base64

i think it's good to implement such helpers, i'll work on it.

**Comment by KonstantinDjairo** [comment]
update:

![Image](https://github.com/user-attachments/assets/e0b29ff4-264a-46fa-9102-b8477f5e8a75)

now that we have the content in base64, it'd reasonable to think of creating more helpers to deal with media.
im going to look into this

**Comment by KonstantinDjairo** [comment]
progress report: (with the help of Robert)

![Image](https://github.com/user-attachments/assets/822e5597-be94-4da5-a690-48c0eca6862a)

i'll finish the function that gives a list of all files in a mdd.
but i have some questions regarding the lookup function; are we able to lookup things by record instead of headword?

**Comment by terasum** [comment]
@KonstantinDjairo yes we can use the record to lookup something, but to building a full records index is a slow process in some  cases.
Actually, the goldendict discard the headword (first/last keyword), and use the records directly.

**Comment by KonstantinDjairo** [comment]
dont worry, the way that i'm planning will be fast.
the only thing i need is to be able to pass the record number instead of the headword, any ideas on how to approach this?

**Comment by terasum** [comment]
@KonstantinDjairo Yes you can use the number which called `meaning_offset`, you can refer this picture: https://github.com/terasum/js-mdict/blob/master/docs/mdict-format.svg , the thing you should resolve is how to get the correct `meaning_offset` number by `headword`.

**Comment by terasum** [comment]
@KonstantinDjairo there also is a reference implement on JavaScript, https://github.com/terasum/js-mdict/blob/1c655578aef121cbc95dbbcac3b5f44b312b1b5d/src/mdict.ts#L83 , FYI

**Comment by KonstantinDjairo** [comment]
thanks for all the information, i'll take a look, for sure.
i first pushed a PR to add the helpers related to MDD data

**Comment by terasum** [comment]
@KonstantinDjairo Ok, Merged just now, thanks for your time! I'm very pleasure to have your help!

**Comment by NiuGuohui** [comment]
hey! There has some errors on arm platform:

<img width="945" alt="Image" src="https://github.com/user-attachments/assets/b8d3df9b-1e6b-430a-9155-980d92f0d9ed" />

**Comment by NiuGuohui** [comment]
I think that in `Base64.hpp`, the content of calling SIMD related instructions can be replaced by simde library. This can be compatible with more platforms

**Comment by terasum** [comment]
@NiuGuohui base64 convert library is a really simple library, you can change into std library for compatition, this won't cost much CPU resource.

**Comment by NiuGuohui** [comment]
> [@NiuGuohui](https://github.com/NiuGuohui) base64 convert library is a really simple library, you can change into std library for compatition, this won't cost much CPU resource.
> 
> Besides base64 part, the others part is compatiable with Apple sillicon chip platform, have fun!

So, our library won't focus on platform compatibility? Do you mean that these modifications for cross-platform compatibility should be implemented by the porting platform (or developers) themselves?

**Comment by terasum** [comment]
@NiuGuohui by the way, there is another libarary meets  platform compatibility, [js-mdict](https://github.com/terasum/js-mdict), FYI.

**Comment by NiuGuohui** [comment]
Thank you. Although I am not using it on the web/nodejs platform, thank you for your reference.

**Comment by terasum** [comment]
@NiuGuohui another solution is checkout the earlier commit before this [PR](https://github.com/dictlab/mdict-cpp/pull/20/), the main function also works, although the mdd file will output as binary format.

try this:
```
git checkout 9c2b73f8c912893e11d6c83f6eb49de4ace901a4
```

**Comment by KonstantinDjairo** [comment]
> I think that in `Base64.hpp`, the content of calling SIMD related instructions can be replaced by simde library. This can be compatible with more platforms

hello.
as by now mdict-cpp has no external dependencies and i think it's better to keep it in that way.

as mentioned by @terasum , running mdict-cpp on a wide branch of architectures isn't a main goal, since the mdict implementation is hard to work with

if you have any idea of how to solve the problem without adding an external dependencies, then PRs are welcome

**Comment by KonstantinDjairo** [comment]
> [@KonstantinDjairo](https://github.com/KonstantinDjairo) Yes you can use the number which called `meaning_offset`, you can refer this picture: https://github.com/terasum/js-mdict/blob/master/docs/mdict-format.svg , the thing you should resolve is how to get the correct `meaning_offset` number by `headword`.

what about [this](https://github.com/dictlab/mdict-cpp/blob/a0a701ad31143ef76cb169fb98584eb38ec95486/mdict.cc#L1463) function?
it seems that it can lookup by record

**Comment by terasum** [comment]
@KonstantinDjairo hi, for compatiable issue, I update the base64 library as turbo-base64 library, but may it's too havy though.

**Comment by KonstantinDjairo** [comment]
> [@KonstantinDjairo](https://github.com/KonstantinDjairo) hi, for compatiable issue, I update the base64 library as turbo-base64 library, but may it's too havy though.

our current base64 library is already bloat-free, so i wont be replacing it

**Comment by KonstantinDjairo** [comment]
> [@KonstantinDjairo](https://github.com/KonstantinDjairo) hi, for compatiable issue, I update the base64 library as turbo-base64 library, but may it's too havy though.

i saw that turbo-base64 is header-only, thanks for considering a self-contained option, because i really like the idea of mdict-cpp being independent and self-contaned

the next thing i want to add is a proper function to search files in the mdd, since we dont output binary anymore... and we have a good decoder that can decode the filenames too, i've added in the base64 PR, it's a hex_to_utf8 decoder

**Comment by terasum** [comment]
@KonstantinDjairo yeah, I recognized that the base64.h contains the other encoding convertion code, I'll revert that.

**Comment by KonstantinDjairo** [comment]
i was thinking, and since mdict-cpp can deliver base64 encoded data instead of binary, it'd be good to add a feature that delivers html pages with embed base64 data, it'd lookup the mdd file and convert images/audio to base64 before sending the final html, this would reduce latency in I/O, at the cost of RAM/memory, but since nowadays people have more than 2gb of ram in their computers, i think it's doable to do it like this

**Comment by terasum** [comment]
Actually, I’ve implements this before, but because of the resource file type are variabous, the html cannot support the all of the content type, like some audio format (*.spx).
At my other repository (github.com/terasum/medict), I implement this function at wapper application layer (wails by golang).


In my opinion, our library inplement the html generation function is too aggresive, maybe some upper layer application take this owner will be better? How do you think? 




---- Replied Message ----
| From | ***@***.***> |
| Date | 05/17/2025 12:40 |
| To | ***@***.***> |
| Cc | Quan ***@***.***> ,
***@***.***> |
| Subject | Re: [dictlab/mdict-cpp] mdd data access? (Issue #13) |
KonstantinDjairo left a comment (dictlab/mdict-cpp#13)

i was thinking, and since mdict-cpp can deliver base64 encoded data instead of binary, it'd be good to add a feature that delivers html pages with embed base64 data, it'd lookup the mdd file and convert images/audio to base64 before sending the final html, this would reduce latency in I/O, at the cost of RAM/memory, but since nowadays people have more than 2gb of ram in their computers, i think it's doable to do it like this

**Comment by KonstantinDjairo** [comment]
maybe we can have a list of supported media, and a fallback function for unsupported media

**Comment by KonstantinDjairo** [comment]
which function do we use to retrieve a dictionary's description?
example of description:

<img src="https://github.com/user-attachments/assets/6210192a-8bb6-4a83-87b0-81485acaaeb9" alt="drawing" width="500"/>

**Comment by terasum** [comment]
@KonstantinDjairo The dictionary's description is saved at `headinfo` 
ref: 
https://github.com/dictlab/mdict-cpp/blob/7162aa232d610ff2985fb77cf539b030ad035971/mdict.cc#L106C1-L106C77 

However we have not a method to read it directly, you can add one :)

**Comment by KonstantinDjairo** [comment]
> [@KonstantinDjairo](https://github.com/KonstantinDjairo) The dictionary's description is saved at `headinfo` ref: https://github.com/dictlab/mdict-cpp/blob/7162aa232d610ff2985fb77cf539b030ad035971/mdict.cc#L106C1-L106C77
> 
> However we have not a method to read it directly, you can add one :)

thank you !
it should be easy to implement i guess

**Comment by terasum** [comment]
the `mdict.locate` function support the mdd data access, and also support read the header file, close this issue fot now.
